jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java

```diff
@@ -589,6 +561,79 @@ else if (authenticationState == null)
         }
     }
 
+    private boolean doWriteError(Request request, Response response, Callback callback, int status)
+    {
+        AuthenticationState authenticationState = AuthenticationState.writeError(request, response, callback, status);
+        if (authenticationState instanceof AuthenticationState.ServeAs serveAs)
+        {
+            RequestResponse result = serveAsWrap(request, response, serveAs);
+            request = result.request;
+            response = result.response;
+            authenticationState = _deferred;
+
+            AuthenticationState.setAuthenticationState(request, authenticationState);
+            IdentityService.Association association =
+                (authenticationState instanceof AuthenticationState.Succeeded user)
+                    ? _identityService.associate(user.getUserIdentity(), null) : null;
+
+            try
+            {
+                //process the request by other handlers
+                return getHandler().handle(_authenticator.prepareRequest(request, authenticationState), response, callback);
+            }
+            catch (Throwable t)
+            {
+                Response.writeError(request, response, callback, HttpStatus.INTERNAL_SERVER_ERROR_500, t.getMessage());
+            }
+            finally
+            {
+                if (association == null && authenticationState instanceof AuthenticationState.Deferred deferred)
+                    association = deferred.getAssociation();
+                if (association != null)
+                    association.close();
+            }
+        }
+
+        return true;
+    }
+
+    private RequestResponse serveAsWrap(Request request, Response response, AuthenticationState.ServeAs serveAs)
```

Comment:
Just return the response and use getRequest

Reply:
done